E2E Tests: Use matrix to run tests with and without GB plugin

ockham opened the pull request:
#### Changes proposed in this Pull Request:

We currently run our e2e tests in two different ways:
- Once for a snapshot of the Jetpack plugin per the current PR.
- Once with the latest Gutenberg plugin installed.

The latter is supposed to help us uncover issues with our own GB blocks that depend on GB, in case the new GB plugin version makes changes that can cause breakage.

There might be opportunity to extend this further, such as also running them with the latest Gutenberg plugin _Release Candidate_ (rather than stable release) installed; or in different browsers (see #17893 for an ongoing experiment).

Anyway, the two ways we're currently running the e2e tests are implemented separately, even though they share most of the logic.

This PR thus attempts to merge them into one single flow, with a [build matrix](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/managing-complex-workflows#using-a-build-matrix) to distinguish the different bits. This should make the GH action more maintainable and extensible.

#### Jetpack product discussion
N/A

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
Check out the relevant e2e tests GH action at the bottom end of this PR, and inspect to verify that e2e tests are still correctly run for both scenarios.

![image](https://user-images.githubusercontent.com/96308/100921313-ab84f900-34dc-11eb-96e8-3a1554a38548.png)

#### Proposed changelog entry for your changes:
E2E Tests: Use matrix to run tests with and without GB plugin

ockham committed:
E2E Tests: Use matrix to run tests with and without GB plugin

ockham committed:
eqeqeq -> eqeq

anomiex commented on .github/workflows/e2e-tests.yml:
```diff
@@ -44,15 +48,21 @@ jobs:
       env:
         NGROK_KEY: ${{ secrets.NGROK_KEY }}
         CONFIG_KEY: ${{ secrets.E2E_CONFIG_KEY }}
+        LATEST_GUTENBERG: ${{ matrix.gutenberg == 'latest' }}
```
This winds up setting LATEST_GUTENBERG to "true" or "false", I believe, while the script looks for empty versus non-empty. See at https://github.com/Automattic/jetpack/pull/17950/checks?check_run_id=1487874826#step:7:28 it's installing Gutenberg anyway.

The most straightforward fix would probably be to change this to just
```suggestion
        GUTENBERG: ${{ matrix.gutenberg }}
```
and have the script do the comparison instead, like
```bash
if [ "$GUTENBERG" == "latest" ]; then
```

There's also another reference to LATEST_GUTENBERG in ./tests/e2e/lib/reporters/slack.js that might also need to be updated.

BTW, does that slack.js reporter get called anywhere below where the environment variable would need to be set? Or if you're unsure you could just move the `env` for it to the job level rather than having it on this one step.

ockham replied:
> This winds up setting LATEST_GUTENBERG to "true" or "false", I believe, while the script looks for empty versus non-empty. See at https://github.com/Automattic/jetpack/pull/17950/checks?check_run_id=1487874826#step:7:28 it's installing Gutenberg anyway.

Ah, great spot, thanks a lot! Sloppy me only checked the with-GB job to see if it's being installed :sweat_smile: 

> The most straightforward fix would probably be to change this to just [...] and have the script do the comparison instead, like
> 
> ```shell
> if [ "$GUTENBERG" == "latest" ]; then
> ```

Yep, sounds good, I'll do that :+1: 

> There's also another reference to LATEST_GUTENBERG in ./tests/e2e/lib/reporters/slack.js that might also need to be updated.

:+1: 

> BTW, does that slack.js reporter get called anywhere below where the environment variable would need to be set? Or if you're unsure you could just move the `env` for it to the job level rather than having it on this one step.

I'll have a look :+1:

ockham replied:
> BTW, does that slack.js reporter get called anywhere below where the environment variable would need to be set? Or if you're unsure you could just move the `env` for it to the job level rather than having it on this one step.

Seems like the slack reporter [is used during global teardown](https://github.com/Automattic/jetpack/blob/07a725146df465cc8df0f08adaa840cd41a4f582/tests/e2e/lib/global-teardown.js#L4) (e2e). I think this would affect at least the with-Gutenberg and without-Gutenberg e2e test steps. Seems like a move to job level would be a good idea :+1: 

(Generally, I'd love if we didn't need to rely on env vars so implicitly, but that's outside of the scope of this PR.)

ockham committed:
Update .github/workflows/e2e-tests.yml

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

ockham committed:
Use GUTENBERG env var rather than LATEST_GUTENBERG

ockham committed:
Move GUTENBERG env var to job level

ockham committed:
Set fail-fast to false

We don't want to e.g. cancel the 'regular' e2e tests if the GB plugin ones fail.

ockham commented:
Similarly to https://github.com/Automattic/jetpack/pull/18024#pullrequestreview-547565829, e2e tests here seem to be consistently failing either with the plugin, or without. I wonder if there's an underlying systematic reason, or if it's just flakiness.

Rebased for good measure.

anomiex commented on .github/workflows/e2e-tests.yml:
```diff
@@ -49,10 +55,15 @@ jobs:
+
     - name: Upload Allure artifacts
-      if: ${{ always() }}
+      if: ${{ matrix.gutenberg == 'no' }}
```
This might be a behavior change.

If I understand [the docs](https://docs.github.com/en/free-pro-team@latest/actions/reference/context-and-expression-syntax-for-github-actions#job-status-check-functions) correctly, the `if: always()` would make this step run even if the tests failed. Changing it to `if: matrix.gutenberg == 'no'` will only run it if the tests passed. If we want to keep the previous behavior, we'd have to do like `if: matrix.gutenberg == 'no' && always()`

ockham replied:
Hmm, I thought that this is the right change: 

Before:

- We ran this job `always` for the without-GB workflow.
- We didn't run it for the with-GB workflow.

Now that we've merged workflows, I've changed the criterion to only run for the without-GB workflow

FWIW, the docs you linked say

> always
>
> Always returns `true`, even when canceled. [...]

That would mean that `matrix.gutenberg == 'no' && always()` is equivalent to `matrix.gutenberg == 'no'` anyway, no?

anomiex replied:
This is the relevant bit:

> You can use the following status check functions as expressions in `if` conditionals. If your `if` expression does not contain any of the status functions it will automatically result with `success()`.

Just `matrix.gutenberg == 'no'` is implicitly equivalent to `matrix.gutenberg == 'no' && success()`, since none of the status functions were included. Adding the `&& always()` avoids the implicit `&& success()`.

anomiex replied:
I did a test at https://github.com/anomiex/testing/pull/1/checks?check_run_id=1539150166. The "If with no status function" step there had `if: ${{ 'a' == 'a' }}` and you can see it skipped running that step. The "If with status function" step had `if: ${{ 'a' == 'a' && always() }}`, and you can see it still ran the step even though the first step in the job failed.

ockham replied:
Updated accordingly :+1:

anomiex commented on tests/e2e/lib/reporters/slack.js:
```diff
@@ -5,14 +5,14 @@ import { readFileSync, createReadStream } from 'fs';
 		this.webCli = new WebClient( token );
 		this.runURL = `https://github.com/Automattic/jetpack/actions/runs/${ GITHUB_RUN_ID }`;
-		this.runType = LATEST_GUTENBERG ? 'with latest :gutenberg:' : 'All';
+		this.runType = GUTENBERG === 'latest' ? 'with latest :gutenberg:' : 'All';
```
Should this match the changes to the job name in the yaml file?

ockham replied:
Uh, match how? Sorry, not seeing/grokking it :sweat_smile:

anomiex replied:
So before we had jobs named "with latest Gutenberg" and "All". This is outputting more or less same thing.

Now our jobs are named "E2E tests: with latest plugin" and "E2E tests: with no plugin". It might be confusing if the slack message still says "All" when it's referring to the "with no plugin" test.

ockham replied:
Ah, gotcha! Yep, I'll change that :+1:

ockham committed:
Update Slack reporter labels

ockham committed:
Restore always uploading Allure artifacts

ockham committed:
Omit redundant expression syntax for if statements

ockham commented:
Addressed feedback, and rebased.

ockham commented:
Thanks for your guidance and reviews, @anomiex!
